node/impl/backupmgr/backup.go

```diff
@@ -0,0 +1,314 @@
+	"boost.db",
+}
+
+type BackupMgr struct {
+	src     lotus_repo.LockedRepo
+	ds      dtypes.MetadataDS
+	sqlConn *sqlite3.SQLiteConn
+	dbName  string
+	lck     sync.Mutex
+}
+
+func NewBackupMgr(src lotus_repo.LockedRepo, ds datastore.Batching, sqlConn *sqlite3.SQLiteConn, name string) *BackupMgr {
+	return &BackupMgr{
+		src:     src,
+		ds:      ds,
+		sqlConn: sqlConn,
+		dbName:  name,
+	}
+}
+
+func (b *BackupMgr) Backup(ctx context.Context, dstDir string) error {
+	_, err := os.Stat(dstDir)
+	if os.IsNotExist(err) {
+		return fmt.Errorf("could not open backup location: %w", err)
+	}
+
+	return b.takeBackup(ctx, dstDir)
+}
+
+func (b *BackupMgr) takeBackup(ctx context.Context, dstDir string) error {
+	s := b.lck.TryLock()
+	if !s {
+		return fmt.Errorf("unable to take lock for backup, please check if there is already another backup running")
+	}
+	defer b.lck.Unlock()
+
+	// Create tmp backup directory and open it as Boost repo
+	bkpDir, err := ioutil.TempDir("", "boost_backup_")
+	if err != nil {
+		return err
+	}
+
+	defer os.RemoveAll(bkpDir)
+
+	dstRepo, err := lotus_repo.NewFS(bkpDir)
+	if err != nil {
+		return err
+	}
+
+	dst, err := dstRepo.Lock(repo.Boost)
+	if err != nil {
+		return err
+	}
+
+	// Create keystore at backup location
+	if err := os.Mkdir(path.Join(dst.Path(), "keystore"), 0700); err != nil {
+		return fmt.Errorf("error creating keystore directory %s: %w", path.Join(dst.Path(), "keystore"), err)
+	}
+
+	// Copy keys from Boost to the backup repo
+	err = CopyKeysBetweenRepos(b.src, dst)
+	if err != nil {
+		return err
+	}
+
+	// Copy metadata from boost repo to backup repo
+	fpath, err := homedir.Expand(path.Join(dst.Path(), MetadataFileName))
+	if err != nil {
+		return fmt.Errorf("expanding metadata file path: %w", err)
+	}
+
+	err = BackupMetadata(ctx, b.ds, fpath)
+	if err != nil {
+		return err
+	}
+
+	// Backup the SQL DBs
+	if err := sqlBackup(b.sqlConn, bkpDir, b.dbName); err != nil {
+		return err
+	}
+
+	// Generate the list of files to be copied from boost repo to the backup repo
+	fl, err := GenerateBkpFileList(b.src.Path(), false)
+	if err != nil {
+		return fmt.Errorf("unable to generate list of files: %w", err)
+	}
+
+	// Copy files
+	if err := os.Mkdir(path.Join(bkpDir, ConfigDirName), 0755); err != nil {
+		return fmt.Errorf("error creating config directory %s: %w", path.Join(bkpDir, ConfigDirName), err)
+	}
+	if err := CopyFiles(b.src.Path(), dst.Path(), fl); err != nil {
+		return fmt.Errorf("error copying file: %w", err)
+	}
+
+	// Close backup repo before moving it
+	if err := dst.Close(); err != nil {
+		return fmt.Errorf("error closing the locked backup repo: %w", err)
+	}
+
+	// Move directory to the backup location
+	if err := os.Rename(bkpDir, path.Join(dstDir, "boost_backup_"+time.Now().Format("20060102150405"))); err != nil {
+		return fmt.Errorf("error moving backup directory %s to %s: %w", bkpDir, dstDir, err)
+	}
+
+	// TODO: Archive the directory and save it to the backup location
```

Comment:
We can remove this TODO right?

Reply:
Yeah, I don't plan to create zip files anymore. At least till folks complaint about it.